### PR TITLE
Don't seek track after analysis if user has seeked it already

### PR DESCRIPTION
**Pull request by daschuer**
This should fix https://bugs.launchpad.net/mixxx/+bug/1879166

### CODE REVIEW

**Comment by ronso0** [comment]
This is an improvement indeed, but it doesn't solve https://bugs.launchpad.net/mixxx/+bug/1879166 for me. I'd be really happy if all Cue automation could be disabled no matter if I seeked or not.

**Comment by daschuer** [comment]
I have probably not fully understand the issue. I hope that everyone with your use case suffers the same issue and we can get around an additional preference option.

You set cue 1 always at 0:00, even if there is a long silent period and the start?
What is you track load position? How do you use the main cue?

**Comment by ronso0** [comment]
I set hotcue 1 at 0:00 (no manual seeking) for 99% of my tracks unless it's one of my recordings with long silence at the beginning. During hotcue preparation I use the main Cue and CUP temporarily to check wether the hotcue is set correctly.
Tracks I'm unsure about what they are and if I want to keep them I just want to play from 0:00 prior to setting any any hotcues.

So my use case is:
* always set Cue at 0:00 while analyzing so I can start the track from zero
* load tracks at Cue

It's simply the silence detection and auto Cue setting (and seeking) that foils my use case.

So I don't really see how to 'repair' my use case and keep it "as is" for others without an Pref option.
Yes, we should not pollute the Prefs with too many options, but I think a
`[ ] enable silence detection and set Cue automatically` checkbox in Pref > Normalization (checked by default) would be valid.

**Comment by daschuer** [comment]
I use also he load at cue setting and I like shifting the cue to intro in and seek the track to this point after analyze, because this way AutoDJ is able to skip the silence.

So it looks to me that shifting is only an issue If you are 
* manually cuing the track during he analysis 
* but you have not jet skeed the track, 
* nor set the Main cue point. 

Is this auto seek actually still an issue? The cue 1 may not go to 0:00  or silent start, depending on the moment you set it. But is this actually bad or more or less a feature? 

If this case is still the issue, do you have an idea how Mixxx can auto-Detect the situation?

**Comment by Be-ing** [comment]
I'm also not understanding what the problem is. If that's how you use hotcues, why do you not want to set the first one at the first sound? What is the advantage of setting it at 0:00 if the track starts with silence?

I have shift + cue mapped to `start_stop` on my controller so I can quickly go back to 0:00 regardless of the position of the cue point or intro start.

**Comment by ronso0** [comment]
I think the issue boils down to: I don't want any automation to interfere with how I cue tracks. The pattern I tried to describe may not always apply, and yet I always want to have total manual control over playposition and cues, and it's also about trust and Mixxx being reliable/predictable and me not having to watch out for "did it seek while I didn't look at it for a second?"
basically "DJ my way!"

I don't mind if it sets intro/outro cues silently as I don't have them enabled anyway.
We can either invest energy trying to catch my use case (which is hard to describe and not always the same), train automation and make the analyze & cue setting more complex and error-prone, probably disturb other use cases by that...or have a simple, distinct option that accomplishes the same and maintains the functionality for others.

**Comment by daschuer** [comment]
@ronso0: I am torn here. 

On one hand you have understandable outlined why you prefer an additional checkbox here.
On the other hand I am very much expecting that you don't notice any issues after this is merged as it is. Because it fixes an issue I have not actually noticed but I will probably suffers as well. 

I was already convinced about the additional checkbox and I have started to implement that, but now I am struggling how to describe it and if it really the seeking that is annoying or the move of the cue point itself and how it relates to the different cue modes.   

How about merging this fix and check out if the remaining issue for for you is really still that annoying that it rectifies a parameter other users not following this discussion will stumble over. 

What do you think? 

At least we can merge this here as a step forward. Right?

**Comment by ronso0** [comment]
> I was already convinced about the additional checkbox and I have started to implement that, but now I am struggling how to describe it and if it really the seeking that is annoying or the move of the cue point itself and how it relates to the different cue modes.

For me, it's both moving the Cue point and the seeking. If both is off it doesn't relate to Cue mode or load behaviour as only unanalyzed tracks are affected.

I will test this on the weekend, thx!

**Comment by daschuer** [comment]
@ronso0: Did you have test results? I suggest to merge this, since it improves the situation and keep the bug open until we have a final solution.

**Comment by ronso0** [comment]
Jup, thank you, it works as expected.
LGTM! (UX-wise)

**Comment by daschuer** [comment]
Cool, thank you.
So I only need to fix the tests.

**Comment by ninomp** [comment]
I tested this PR and noticed an issue. Here are steps to reproduce:
1. In Beat Detection preferences enable reanalysis
2. In Decks preferences set Track load point to Intro start
3. Turn quantization on
4. Load a track that has out-of-date beatmap
5. When analysis finishes, intro start is moved due to the updated beatmap and track isn't seeked to follow the updated intro start

Expected behavour: Upon completion of analysis, track is seeked to the new position of intro start.

File: src/test/cuecontrol_test.cpp [hunk 1]
```diff
@@ -266,35 +266,53 @@ TEST_F(CueControlTest, SeekOnLoadDefault) {
...
             ConfigValue(static_cast<int>(SeekOnLoadMode::MainCue)));
     TrackPointer pTrack = createTestTrack();
-    pTrack->setCuePoint(CuePosition(-100.0));
-
     loadTrack(pTrack);
 
-    EXPECT_DOUBLE_EQ(-100.0, m_pCuePoint->get());
-    EXPECT_DOUBLE_EQ(-100.0, getCurrentSample());
+    EXPECT_DOUBLE_EQ(0.0, m_pCuePoint->get());
+    EXPECT_DOUBLE_EQ(0.0, getCurrentSample());
```
Comment (ninomp) [inline src/test/cuecontrol_test.cpp:315]: Why is this changed from `-100` to `0`? The purpose of this test was to verify that track is seeked to main cue when it is set in preroll (`cuePosition < 0`). This was a bug previously: https://bugs.launchpad.net/mixxx/+bug/1853919

File: src/test/cuecontrol_test.cpp [hunk 2]
```diff
@@ -290,29 +290,33 @@ TEST_F(CueControlTest, DontSeekOnLoadMainCue) {
     TrackPointer pTrack = createTestTrack();
     pTrack->setCuePoint(CuePosition(100.0));
 
-    // The Track should not follow if the cue point has been set manually in between.
+    // The Track should not follow if the track has been manual set manually in between.
```
Comment (Holzhaus) [inline src/test/cuecontrol_test.cpp:293]: I don't understand this comment. Can you rephrase?

**Comment by daschuer** [comment]
Done.

**Comment by ninomp** [comment]
LGTM

File: src/engine/controls/cuecontrol.h [hunk 3]
```diff
@@ -203,6 +203,7 @@ class CueControl : public EngineControl {
     ControlObject* m_pQuantizeEnabled;
     ControlObject* m_pClosestBeat;
     bool m_bypassCueSetByPlay;
+    ControlValueAtomic<double> m_oldSeekOnLoadPosition;
```
Comment (uklotzde) [inline src/engine/controls/cuecontrol.h:206]: Why "m_oldSeekOnLoadPosition" instead of just "m_seekOnLoadPosition"? Otherwise I would expect that there also is a *current* or *new* position.

**Review by uklotzde — CHANGES_REQUESTED** [review]

File: src/engine/controls/cuecontrol.cpp [hunk 4]
```diff
@@ -356,6 +356,7 @@ void CueControl::trackLoaded(TrackPointer pNewTrack) {
         m_pOutroEndEnabled->forceSet(0.0);
         m_pHotcueFocus->set(Cue::kNoHotCue);
         m_pLoadedTrack.reset();
+        m_oldSeekOnLoadPosition.setValue(0.0);
```
Comment (uklotzde) [inline src/engine/controls/cuecontrol.cpp:359]: kDefaultSeekOnLoadPosition?

File: src/engine/controls/cuecontrol.h (hunk 3 above)
Comment (daschuer) [reply src/engine/controls/cuecontrol.h:206]: It is like that. "m_oldSeekOnLoadPosition" is the position that has been taken because of the old cue points before the analysis has finished. The new position is calculated on the fly and copied into old. "m_seekOnLoadPosition" does not work for me for the case. Do you have a better fitting name?

File: src/engine/controls/cuecontrol.cpp (hunk 4 above)
Comment (daschuer) [reply src/engine/controls/cuecontrol.cpp:359]: This is more like the uninitialized value "kNotSeekedOnLoadPosition". This should fit to the new name for the variable we decide for. Ideas?

File: src/engine/controls/cuecontrol.h (hunk 3 above)
Comment (uklotzde) [reply src/engine/controls/cuecontrol.h:206]: You explained that m_oldSeekOnLoadPosition always reflects the current *seek-on-load* position that is valid at that point in time. After the analysis has finished it is replaced if not already set to a valid position. So why old? It does never get *old* since it is updated.

File: src/engine/controls/cuecontrol.cpp (hunk 4 above)
Comment (uklotzde) [reply src/engine/controls/cuecontrol.cpp:359]: No need to argue about names. But we use 0.0 too often with lots of implicit meanings.

File: src/engine/controls/cuecontrol.h (hunk 3 above)
Comment (daschuer) [reply src/engine/controls/cuecontrol.h:206]: The only propose is to compare the old position with the new one. In such cases I use always the term old. 
This way it is clear "don't use the value again, it is old".

**Comment by daschuer** [comment]
I have refactored the code a bit to be less confusing.

File: src/engine/controls/cuecontrol.cpp [hunk 5]
```diff
@@ -432,27 +440,33 @@ void CueControl::trackLoaded(TrackPointer pNewTrack) {
...
-        seekExact(0.0);
+        DEBUG_ASSERT(!"Unknown enum value");
+        seekOnLoad(0.0);
         break;
     }
 }
 
+void CueControl::seekOnLoad(double seekOnLoadPosition) {
+    seekExact(seekOnLoadPosition);
+    m_usedSeekOnLoadPosition.setValue(seekOnLoadPosition);
```
Comment (uklotzde) [inline src/engine/controls/cuecontrol.cpp:467]: Why not `m_lastSeekOnLoadPosition`?

Comment (daschuer) [reply src/engine/controls/cuecontrol.cpp:467]: Because it is only relevant if the last seek position has been used. 
The last position is not used in case of vinyl control.

**Comment by daschuer** [comment]
Can we merge this now?

**Review by uklotzde — APPROVED** [review]

**Comment by uklotzde** [comment]
LGTM

**Comment by daschuer** [comment]
Cool, this has also fixed https://bugs.launchpad.net/mixxx/+bug/1888012
